### PR TITLE
Core/Combat: Cleanup DoZoneInCombat function.

### PR DESCRIPTION
**Pull request by UltraNix**
Updates #7960

## Tests Performed:

- Tested with many bosses/scripts where `DoZoneInCombat` is called, e.g. Trial of the Crusader.

## How to Test the Changes:

1. Lookup the script where you can see `DoZoneInCombat` call and trigger it.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by UltraNix** [comment]
@patou01 Check now.

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -9715,7 +9715,7 @@ bool Unit::Attack(Unit* victim, bool meleeAttack)
     // set position before any AI calls/assistance
     //if (GetTypeId() == TYPEID_UNIT)
     //    ToCreature()->SetCombatStartPosition(GetPositionX(), GetPositionY(), GetPositionZ());
-    if (creature && !IsControlledByPlayer())
+    if (creature && !(IsControllableGuardian() && IsControlledByPlayer()))
```
Comment (Nyeriah) [inline src/server/game/Entities/Unit/Unit.cpp:9718]: I believe this part got merged in accidentally

File: src/server/game/Entities/Creature/Creature.cpp [hunk 2]
```diff
@@ -2550,40 +2550,8 @@ void Creature::SendZoneUnderAttackMessage(Player* attacker)
...
-
-            if (player->IsAlive())
-            {
-                this->SetInCombatWith(player);
-                player->SetInCombatWith(this);
-                AddThreat(player, 0.0f);
-            }
-        }
-    }
+    if (IsAIEnabled)
```
Comment (patou01) [inline src/server/game/Entities/Creature/Creature.cpp:2553]: Hmm so a creature that is not AI enabled cannot use doZoneInCombat anymore (well, SetInCombatWithZone). The question is... does anyone do that?

Comment (UltraNix) [reply src/server/game/Entities/Creature/Creature.cpp:2553]: All creatures but charmed have `IsAIEnabled` set to true.

File: src/server/game/AI/CreatureAI.cpp [hunk 3]
```diff
@@ -56,60 +56,74 @@ void CreatureAI::Talk(uint8 id, WorldObject const* whisperTarget /*= nullptr*/)
     sCreatureTextMgr->SendChat(me, id, whisperTarget);
 }
 
-void CreatureAI::DoZoneInCombat(Creature* creature /*= nullptr*/, float maxRangeToNearestTarget /* = 50.0f*/)
+inline bool IsValidCombatTarget(Creature* source, Player* target)
 {
```
Comment (patou01) [inline src/server/game/AI/CreatureAI.cpp:60]: Does it do anything significantly different from CanCreatureAttack() ? (with setting skipDistCheck to false)
image

Comment (UltraNix) [reply src/server/game/AI/CreatureAI.cpp:60]: Yes - it's different. `CanCreatureAttack` checks more things that are not needed here.

File: src/server/game/AI/CreatureAI.cpp [hunk 4]
```diff
@@ -56,60 +56,74 @@ void CreatureAI::Talk(uint8 id, WorldObject const* whisperTarget /*= nullptr*/)
...
+        return false;
+    }
+
+    return true;
+}
+
+void CreatureAI::DoZoneInCombat(Creature* creature /*= nullptr*/, float maxRangeToNearestTarget /*= 250.0f*/)
+{
+    if (!creature)
+        creature = me;
```
Comment (Nefertumm) [inline src/server/game/AI/CreatureAI.cpp:92]: brackets

File: src/server/game/AI/CreatureAI.cpp [hunk 5]
```diff
@@ -56,60 +56,74 @@ void CreatureAI::Talk(uint8 id, WorldObject const* whisperTarget /*= nullptr*/)
...
+
+    if (creature->IsInEvadeMode())
+        return;
```
Comment (Nefertumm) [inline src/server/game/AI/CreatureAI.cpp:95]: same

**Review by Nefertumm — APPROVED** [review]

**Comment by UltraNix** [comment]
@patou01 Have you tested it?

**Comment by patou01** [comment]
Trying it now (well, once my computer is done with compiling)

**Comment by patou01** [comment]
Seems to work for me. Put an alt at distance 110 from the spawnpoint of Urok, he didn't get put in combat.  No warning when I put my gm tag on (ie nobody for him to aggro).

**Pull request merged**